src/cascadia/Remoting/Monarch.h

```diff
@@ -69,14 +70,19 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
     private:
         uint64_t _ourPID;
 
-        uint64_t _nextPeasantID{ 1 };
-        uint64_t _thisPeasantID{ 0 };
+        std::atomic<uint64_t> _nextPeasantID{ 1 };
+        uint64_t _ourPeasantId{ 0 };
+
+        // When we're quitting we do not care as much about handling some events that we know will be triggered
+        std::atomic<bool> _quitting{ false };
 
         winrt::com_ptr<IVirtualDesktopManager> _desktopManager{ nullptr };
 
         std::unordered_map<uint64_t, winrt::Microsoft::Terminal::Remoting::IPeasant> _peasants;
+        std::shared_mutex _peasantsMutex{};
 
         std::vector<Remoting::WindowActivatedArgs> _mruPeasants;
+        std::recursive_mutex _mruPeasantsMutex{};

```

Comment:
_getMostRecentPeasantID calls (through a couple hops) _clearOldMruEntries

Reply:
Oh right... That's unfortunate - recursive mutexes are really annoying (it uses critical sections brrrr).
I wonder if it would simplify the code if we just slapped a single mutex to the beginning of all public methods of this class and made all private methods non-thread-safe. It would get rid of the recursion and the atomics at least... The Monarch class has an extremely slow "throughput" (API calls per second) right?
Hmm... 🤔

Reply:
In normal circumstances, yes. The problem is with loading / quitting where there are potentially "many" (O(#windows)) calls, and of course that is where I found these problems in the first place.

Reply:
I can take a look if there is an easy refactor to get rid of the recursive mutex for a shared_mutex, but I think it is still better to have the more granular locking in general.

Reply:
Without rewriting `_getMostRecentPeasantID` and `_getPeasant` the only safe thing I can think of is to pass a `std::unique_lock<...>*` to `_getPeasant` and `_clearOldMruEntries` as an argument and if that is null then lock in `_clearOldMruEntries`. I'm not a huge fan of that approach because it is error prone, but functional.

Reply:
I made `_getPeasant` take an optional `clearMruPeasantOnFailure` boolean, which is set to false in `_getMostRecentPeasantID`. I then updated `_getMostRecentPeasantID` to handle its own cleanup, and switched the recursive mutex to a shared mutex.

Reply:
Spent more time trying to make sure we can't deadlock. I am ~90% confident, but have not proved, that the current version of the code cannot deadlock. Currently `_getMostRecentPeasantID` (and transitively anything that calls it) is the only method that attempts to hold both locks at the same time. There should not be the case where someone else holds a unique lock on `_peasantsMutex` and is waiting to get a lock on `_mruPeasantsMutex` which would deadlock with `_getMostRecentPeasantID`.